File: backend/medshield_backend/scan/schemas.py
"""Scan request / response models.

Verdict shape mirrors what the Chrome extension consumes - see
``docs/plans/backend.md`` §4 and ``docs/plans/frontend.md`` §6.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, EmailStr, Field, field_validator

Action = Literal["allow", "warn", "block", "quarantine", "escalate"]
Severity = Literal["none", "low", "medium", "high", "critical"]


class EntityHit(BaseModel):
    type: str
    masked_value: str  # never the raw PHI
    confidence: float = Field(ge=0.0, le=1.0)
    source: Literal["body", "subject", "attachment"]
    attachment_id: str | None = None


class RecipientHit(BaseModel):
    email: EmailStr
    classification: Literal[
        "internal",
        "approved_partner",
        "external",
        "public_email",
        "unknown",
    ]


class ScanEmailRequest(BaseModel):
    message_id: str | None = None
    subject: str = ""
    body: str = ""
    recipients: list[EmailStr] = Field(min_length=1)


class ScanFinalizeRequest(BaseModel):
    message_id: str | None = None
    attachment_scan_ids: list[str] = Field(default_factory=list)
    override_quarantine: bool = False


class AttachmentScanResponse(BaseModel):
    scan_id: str
    status: Literal["scanned", "queued"]
    filename: str
    size_bytes: int
    mime_type: str


class Verdict(BaseModel):
    scan_id: str
    action: Action
    severity: Severity
    risk_score: float
    matched_policy_ids: list[str]
    entities: list[EntityHit]
    recipients: list[RecipientHit]
    user_message: str
    created_at: datetime

    @field_validator("risk_score")
    @classmethod
    def round_score(cls, value: float) -> float:
        return round(value, 2)


class ScanStatusResponse(BaseModel):
    scan_id: str
    status: Literal["pending", "scanning", "completed", "failed"]
    verdict: Verdict | None = None

File: frontend/packages/shared/src/schemas.ts
import { z } from 'zod';

export const actionSchema = z.enum(['allow', 'warn', 'block', 'quarantine', 'escalate']);
export const severitySchema = z.enum(['none', 'low', 'medium', 'high', 'critical']);
export const recipientClassSchema = z.enum([
  'internal',
  'approved_partner',
  'external',
  'public_email',
  'unknown',
]);

export const entityHitSchema = z.object({
  type: z.string(),
  masked_value: z.string(),
  confidence: z.number().min(0).max(1),
  source: z.enum(['body', 'subject', 'attachment']),
  attachment_id: z.string().optional(),
});

export const recipientHitSchema = z.object({
  email: z.string().email(),
  classification: recipientClassSchema,
});

export const verdictSchema = z.object({
  scan_id: z.string(),
  action: actionSchema,
  severity: severitySchema,
  risk_score: z.number(),
  matched_policy_ids: z.array(z.string()),
  entities: z.array(entityHitSchema),
  recipients: z.array(recipientHitSchema),
  user_message: z.string(),
  created_at: z.string(),
  quarantine_id: z.string().nullable().optional(),
  degraded: z.boolean().optional(),
});

export const attachmentUploadResultSchema = z.object({
  attachment_scan_id: z.string(),
  status: z.enum(['scanned', 'queued', 'failed']),
  verdict: verdictSchema.nullable().optional(),
  error: z.string().nullable().optional(),
});

export const authTokensSchema = z.object({
  access_token: z.string(),
  expires_in: z.number(),
  token_type: z.literal('Bearer'),
});

export const userProfileSchema = z.object({
  user_id: z.string(),
  email: z.string().email(),
  name: z.string(),
  workspace_id: z.string(),
  role: z.enum(['user', 'analyst', 'admin', 'super_admin']),
});
